src/core/blueprint/into_schema.rs

```diff
@@ -34,14 +38,120 @@ fn to_type_ref(type_of: &Type) -> dynamic::TypeRef {
     }
 }
 
-fn to_type(def: &Definition) -> dynamic::Type {
+/// Entity that analyzes resolved data and decides what variant type of union
+/// this data satisfies to. It's required to specify exact type in
+/// `async_graphql` otherwise validation error will be generated at runtime
+#[derive(Clone)]
+struct UnionTypeResolver {
+    type_name: String,
+    schemas: Arc<Option<Vec<(String, JsonSchema)>>>,
+}
+
+impl UnionTypeResolver {
```

Comment:
Let's move this to `union_resolver.rs`

Reply:
I have a different proposal for this implementation. Ideally we would identify and store the discriminator while converting to blueprint. If we can't find the discriminator for a union type, we should throw a validation error, stating that there is no possible way to discriminate between the variants of the given union type.


Once we identify the discriminator we should create new IR:


```rust
enum IR {
  // ...
  Discriminate(Discriminator, Box<IR>)
  // ...
}

struct Discriminator(Vec<FieldName, TypeName>)

```

This particular IR when evaluated will check to identify the type of the output of the inner IR and store it in EvaluationCtx. In `into_schema` we can just read it and set it in the field value.

Let's discuss on discord if it isn't clear.

I feel this approach will be significantly superior to all existing Union type implementations.